feat: use bound profile provider to call perform.

Jakub-Vacek opened the pull request:
<!--- Tip: You don't have to remove these comments -->

## Description
This PR changes approach to calling SDK. BoundProfileProvider instance is created and used to call perform.
Creation of instance can be divided into two steps:
- resolution of "SF files" we load super.json and local profileAST, mapAST and providerJson (resolveSuperfaceFiles function)
- creation of BoundProfileProvider instance from "SF files"  (createBoundProfileProvider function)

these steps (functions) are used in `prepareSuperface` (naming sucks). This function prepares everything we need to call `perform`. It is easy to mock with `mockBoundProfileProvider` eg. in temporary test file `superface-test-new.test`

This approach can be extended to some (`startRecording` and `endRecording`) functions that are now private in `SuperfaceTest`. That would allow us easier testing (easy mocking + easy unit testing of recording fns)

This approach is, I think:
- easy to test
- easy to read
- unable to use remote files (for now)
- possibly problematic due to OneSdk exports (we need function for preparing BoundProfileProvider exported)

There is a lot of TODOs and thing that can be improved (errors). My main goal was to simplify whole solution and enable reasonable testing.

## Motivation and Context
Using Client that actually calls bind is imo not good approach - it is hard to test and to read. It is also more complicated that it has to be (I believe).

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- If your changes are only to the internals then make sure any function has its attached documentation updated or added (e.g. descriptive name, doc comment, etc.). If your changes are to user-facing APIs or concepts then make sure README.md is updated accordingly (e.g. command help output). -->
- [ ] I have updated the documentation accordingly. If you are changing **code related to user secrets** you need to really make sure that [security documentation](SECURITY.md) is correct.
- [ ] I have read the **CONTRIBUTION_GUIDE** document.
- [ ] I haven't repeated the code. (DRY)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

martinalbert committed:
chore: commit save point

martinalbert committed:
feat: update sdk and ast and resolve breaking changes

martinalbert committed:
chore: update one-sdk to 2.0.0-beta.0

martinalbert committed:
chore: commit save point

martinalbert committed:
feat(client): move TestClient into superface folder

martinalbert committed:
feat(config): wrap superface components TestConfig class
- components such as profile, provider, usecase, boundProfileProvider

martinalbert committed:
feat(superface-test): update class SuperfaceTest
- remove functions for preparing superface components which are moved to Config class

martinalbert committed:
wip: test: try to fix tests
- mocking is not compatible, needs more love

martinalbert committed:
test: skip not passing tests due to problem with mocking

martinalbert committed:
chore: format

martinalbert committed:
wip: feat: update library to support sdk 2.0

Jakub-Vacek committed:
feat: use bound profile provider to call perform.

freaz commented on src/superface-test-new.test.ts:
```diff
@@ -0,0 +1,135 @@
+  let superfaceTest: SuperfaceTest;
+
+  afterEach(() => {
+    jest.restoreAllMocks();
```
suggestion: Resetting mock in afterEach is here in every (I opened) test, so much be better to turn on mock reseting in global jest configuration. https://jestjs.io/docs/configuration#resetmocks-boolean

martinalbert committed:
feat: use SecurityConfigurations

martinalbert committed:
feat: move start and end recording to separate functions

martinalbert committed:
feat: update superface setup functions
- add new error for providerJson not found
- add options with possible mocked parts for boundProfileProvider
- restructure resolving of superface files

martinalbert committed:
feat: remove config and add prepare modules

freaz reviewed (commented):
I like where it is heading with all the changes. @Jakub-Vacek kudos for it.

martinalbert committed:
feat: get prepare module ready for testing

martinalbert committed:
test: add tests for prepare module

martinalbert committed:
chore(deps): move parser to dev and peer deps

martinalbert committed:
test: remove unnecessary tests and code for mocking

martinalbert committed:
test: fix tests

martinalbert committed:
chore(deps): add one-sdk 2.0.0-rc.1

martinalbert committed:
feat: fix tests and lint after sdk upgrade

martinalbert committed:
chore: update fixtures

martinalbert commented:
@Jakub-Vacek @freaz I finished most of what we talked about, so please review 🙏🏻 

I only have to fix not passing tests and maybe figure out how to mock recording better, now that it's abstracted to functions.

martinalbert committed:
fix: repaire not passing tests and lint errors

Jakub-Vacek commented on src/common/format.ts:
```diff
@@ -56,34 +52,37 @@ function isValidPayload(payload: string, match: string): boolean {
+  // sfConfig: Pick<
+  //   CompleteSuperfaceTestConfig,
+  //   'profile' | 'provider' | 'useCase'
+  // >,
```
Remove unused code that I probably left there (sorry) :/

Jakub-Vacek commented on src/superface-test.interfaces.ts:
```diff
@@ -9,8 +8,6 @@ import {
 
-import { ISuperfaceClient } from './superface/client';
-
 export interface SuperfaceTestConfigPayload {
```
I would named this SuperfaceTestConfig. Payload feels unnecessary.

Jakub-Vacek commented on src/superface-test.test.ts:
```diff
@@ -1,47 +1,31 @@
 import { ApiKeyPlacement, SecurityType } from '@superfaceai/ast';
```
Tests looks much better now 👍

Jakub-Vacek commented on src/superface-test.test.ts:
```diff
@@ -510,6 +570,27 @@ describe.skip('SuperfaceTest', () => {
         mocked(exists).mockResolvedValueOnce(true);
         mocked(readFileQuiet).mockResolvedValueOnce('[]');
         mocked(matchWildCard).mockReturnValueOnce(false);
+        mocked(prepareSuperface).mockResolvedValue({
```
mocking of prepareSuperface can be probably move to some top level function to get rid of repeating it in every case.

Jakub-Vacek commented on src/superface/config/create-bound-profile-provider.ts:
```diff
@@ -0,0 +1,100 @@
+  resolveSecurityConfiguration,
+  ServiceSelector,
+} from '@superfaceai/one-sdk';
+// TODO: export from SDK?
```
You can remove the TODO :)

Jakub-Vacek commented on src/superface/config/prepare-ast.ts:
```diff
@@ -0,0 +1,71 @@
+
+    // TODO: resolve path
+    const profilePath = profileSettings.file;
+    const content = await fileSystem.readFile(profilePath);
```
This TODO means that something like this:

resolvePath(dirname(superJsonPath), path)

should be used to resolve the file. Try it in station just to be sure. I am not saying that you need to change just make sure it works :)

martinalbert replied:
After testing in station, I figured out that `dirname` is not necessary as `detectSuperJson` returns path to directory, not the exact path, so I used `resolvePath(superJsonPath, path)`

Jakub-Vacek commented on src/superface/config/prepare-ast.ts:
```diff
@@ -0,0 +1,71 @@
+      throw content.error;
+    }
+
+    // TODO: make parser optional
```
Remove TODO

martinalbert replied:
I added this, I can add dynamic import and call the fallback without parser if import fails. The question is, what will be fallback. 😄

Jakub-Vacek replied:
Fall should be error saying that you need to compile and there also should be extension check. But I am ok with keeping parser in deps.

Jakub-Vacek commented on src/superface/config/prepare-ast.ts:
```diff
@@ -0,0 +1,71 @@
+  }
+}
+
+export async function getMapAst(
```
Same stuff as with profile

Jakub-Vacek commented on src/superface/config/prepare-provider-json.ts:
```diff
@@ -0,0 +1,32 @@
+    throw new ProviderJsonUndefinedError(providerName);
+  }
+
+  // superJson.resolvePath(providerSettings.file);
```
Same as with map :)

Jakub-Vacek reviewed (commented):
Added some comments on the small stuff. Overall looks good.

martinalbert committed:
fix: review fixes

martinalbert committed:
fix: review fixes

martinalbert committed:
feat: move getSuperJson among prepare modules

Jakub-Vacek commented on src/superface/mock/timers.ts:
```diff
@@ -0,0 +1,48 @@
+  constructor(public value: number) {}
+}
+
+export class MockTimers implements ITimers {
```
Why didn't you just reused mocks from SDK?

martinalbert replied:
Because it's not exported from sdk as far as I know 😄

martinalbert committed:
chore: change version to avoid problems in CI
- from 2.0.3 to 3.0.0-beta.0

Commit:
chore: release v3.0.0-beta.1

martinalbert committed:
fix: add function for resolving credentials

Commit:
chore: release v3.0.0-beta.2

martinalbert committed:
feat: use NodeEnvironment instead of new function

Commit:
chore: release v3.0.0-beta.3

martinalbert committed:
feat: check for ast before parsing

Commit:
chore: release v3.0.0-beta.4

martinalbert committed:
Merge pull request #37 from superfaceai/feat/resolve-credentials

fix: Add function for resolving credentials

Jakub-Vacek commented:
It looks ok (I checked files that were changed after my last review). Please resolve conflicts before final review.

martinalbert committed:
Merge branch 'dev' into feat/use-bound-profile-provider

martinalbert committed:
chore(peerDeps): update version of parser

martinalbert committed:
chore: update changelog
